datasource bigip_ltm_policy is not partition aware

<!--
When filing an issue please check to see if an issue already exists that matches your's issue.

Please open a case (https://support.f5.com/csp/article/K2633) with F5 if this is a critical issue.

-->

### Environment

 * TMOS/Bigip Version: 17.1.2.1
 * Terraform Version: 1.12.1
 * Terraform bigip provider Version: 1.23.1

### Summary

Trying to read existing LTM policy, which is created in partition (partition name is "kube" in this concrete case) fails.

In the error message it seems, that the provided policy name is not interpreted correctly. "/Common/" is prepended and the very end part of the policy name is trimmed (see steps to reproduce)

In the documentation is stated, that the name should start with the partition, and there is no dedicated parameter for the parttion (which is the case, there is dedicated parameter for the partition on sthe imilar datasources, such as bigip_ltm_node, bigip_ltm_irule etc..) 

### Steps To Reproduce

Steps to reproduce the behavior:

1. use this datasoure fragment (make sure it is pointing to existing policy in a partition)
```
data "bigip_ltm_policy" "ccc" {
  name = "/kube/Shared/catcher-ruller"
}
```

2. To get to know more about the issue, provide terraform debug logs
```
TF_LOG=ERROR terraform plan 
data.bigip_ltm_policy.ccc: Reading...
bigip_ltm_virtual_server.vhoster: Refreshing state... [id=/kube/virtualhost-catch]
bigip_ltm_profile_client_ssl.pvnetbox: Refreshing state... [id=/kube/Shared/cather-pvnetbox]
bigip_ltm_profile_client_ssl.netbox: Refreshing state... [id=/kube/Shared/cather-netbox]
bigip_ltm_virtual_server.terra-emptier: Refreshing state... [id=/kube/terra-empty]
bigip_ltm_virtual_server.vs-catch["pvnetbox"]: Refreshing state... [id=/kube/Shared/pvnetbox-catch]
bigip_ltm_virtual_server.vs-catch["oddo"]: Refreshing state... [id=/kube/Shared/oddo-catch]
bigip_ltm_virtual_server.vs-catch["netbox"]: Refreshing state... [id=/kube/Shared/netbox-catch]
bigip_ltm_profile_client_ssl.odoo: Refreshing state... [id=/kube/Shared/cather-odoo]
bigip_ltm_policy.catcher-ruller: Refreshing state... [id=/kube/Shared/catcher-ruller]
2025-07-15T10:27:15.612+0300 [ERROR] provider.terraform-provider-bigip_v1.23.1: Unable to Retrieve Policy   (Shared) (01020036:3: The requested Policy (/Common/kube/Shared) was not found.): timestamp="2025-07-15T10:27:15.611+0300"
2025-07-15T10:27:15.612+0300 [ERROR] provider.terraform-provider-bigip_v1.23.1: Response contains error diagnostic: diagnostic_summary="01020036:3: The requested Policy (/Common/kube/Shared) was not found." tf_data_source_type=bigip_ltm_policy tf_proto_version=5.3 tf_req_id=229a268f-8244-4a42-8ed8-d1deaec70c60 tf_rpc=ReadDataSource @caller=github.com/hashicorp/terraform-plugin-go@v0.14.3/tfprotov5/internal/diag/diagnostics.go:55 @module=sdk.proto diagnostic_detail="" tf_provider_addr=provider diagnostic_severity=ERROR timestamp="2025-07-15T10:27:15.612+0300"
2025-07-15T10:27:15.612+0300 [ERROR] vertex "data.bigip_ltm_policy.ccc" error: 01020036:3: The requested Policy (/Common/kube/Shared) was not found.
2025-07-15T10:27:15.612+0300 [ERROR] vertex "data.bigip_ltm_policy.ccc (expand)" error: 01020036:3: The requested Policy (/Common/kube/Shared) was not found.
bigip_ltm_virtual_server.vs-catch["bnra"]: Refreshing state... [id=/kube/Shared/bnra-catch]
data.bigip_ltm_policy.ccc: Still reading... [00m10s elapsed]

Planning failed. Terraform encountered an error while generating this plan.

╷
│ Error: 01020036:3: The requested Policy (/Common/kube/Shared) was not found.
│ 
│   with data.bigip_ltm_policy.ccc,
│   on getpol.tf line 1, in data "bigip_ltm_policy" "ccc":
│    1: data "bigip_ltm_policy" "ccc" {
│ 
╵

```

### Expected Behavior
Expected is the existing ltm policy ( "/kube/Shared/catcher-ruller" in this case ) to be read

### Actual Behavior
The terraform provider (tried with older version too - till 1.21.0) does not honor correctly the provided name of the policy, especially in the term of figuring out the correct partition, and is prepending /Common/, and also trimming the very end of policy name

--- a/bigip/datasource_bigip_ltm_policy.go
+++ b/bigip/datasource_bigip_ltm_policy.go
@@ -1080,21 +1080,27 @@ func dataSourceBigipLtmPolicyRead(ctx context.Context, d *schema.ResourceData, m
 
 	d.SetId("")
 	name := d.Get("name").(string)
-	log.Println("[DEBUG] Reading policy : " + name)
 
-	re := regexp.MustCompile("/([a-zA-z0-9? ,_-]+)/([a-zA-z0-9? ,._-]+)")
-	match := re.FindStringSubmatch(name)
-	if match == nil {
-		return diag.FromErr(fmt.Errorf("policy name failed to match the regex, and should be of format /partition/policy_name"))
+	fields, diags := parsePolicyPath(name)
+	if diags != nil {
+		return diag.FromErr(fmt.Errorf("failed to parse policy path: %v", diags))
 	}
-	partition := match[1]
-	policyName := match[2]
+	log.Printf("[INFO] Fetching policy: %s (partition: %s)", fields[len(fields)-1], fields[0])
 
-	log.Println("[INFO] Fetching policy " + policyName)
-	p, err := client.GetPolicy(policyName, partition)
+	var (
+		p   *bigip.Policy
+		err error
+	)
+	switch len(fields) {
+	case 2:
+		p, err = client.FetchPolicy(fields[0], fields[1])
+	case 3:
+		p, err = client.FetchPolicy(fields[0], fields[1], fields[2])
+	default:
+		return diag.FromErr(fmt.Errorf("unexpected number of path fields: %v", fields))
+	}
 
 	if err != nil {
-		log.Printf("[ERROR] Unable to Retrieve Policy   (%s) (%v) ", policyName, err)
 		return diag.FromErr(err)
 	}
 
@@ -1124,6 +1130,27 @@ func dataSourceBigipLtmPolicyRead(ctx context.Context, d *schema.ResourceData, m
 	}
 	d.SetId(name)
 	return DatapolicyToData(p, d)
+}
+
+// parsePolicyPath parses a policy path in the format /partition[/strategy]/policy_name
+// Returns the field segments or an error diagnostic.
+func parsePolicyPath(name string) ([]string, diag.Diagnostics) {
+	log.Println("[DEBUG] Reading policy:", name)
+
+	// Regex: /partition/(optional strategy/)?policy_name
+	re := regexp.MustCompile(`^/([^/]+)(?:/([^/]+))?/([^/]+)$`)
+	match := re.FindStringSubmatch(name)
+	if match == nil {
+		return nil, diag.FromErr(fmt.Errorf(
+			"policy name failed to match required format: /partition[/strategy]/policy_name, got: %q", name))
+	}
+
+	fields := []string{match[1]}
+	if match[2] != "" {
+		fields = append(fields, match[2])
+	}
+	fields = append(fields, match[3])
+	return fields, nil
 }
 
 func DatapolicyToData(p *bigip.Policy, d *schema.ResourceData) diag.Diagnostics {
@@ -1164,10 +1191,10 @@ func DatapolicyToData(p *bigip.Policy, d *schema.ResourceData) diag.Diagnostics 
 	return nil
 }
 
-func DataflattenPolicyRules(rules []bigip.PolicyRule) []interface{} {
-	att := make([]interface{}, len(rules))
+func DataflattenPolicyRules(rules []bigip.PolicyRule) []any {
+	att := make([]any, len(rules))
 	for i, v := range rules {
-		obj := make(map[string]interface{})
+		obj := make(map[string]any)
 
 		if v.Name != "" {
 			obj["name"] = v.Name
@@ -1188,24 +1215,24 @@ func DataflattenPolicyRules(rules []bigip.PolicyRule) []interface{} {
 	return att
 }
 
-func DataflattenPolicyRuleActions(actions []bigip.PolicyRuleAction) []interface{} {
-	att := make([]interface{}, len(actions))
+func DataflattenPolicyRuleActions(actions []bigip.PolicyRuleAction) []any {
+	att := make([]any, len(actions))
 	for x, a := range actions {
 		att[x] = DatainterfaceToResourceData(a)
 	}
 	return att
 }
 
-func DataflattenPolicyRuleConditions(conditions []bigip.PolicyRuleCondition) []interface{} {
-	att := make([]interface{}, len(conditions))
+func DataflattenPolicyRuleConditions(conditions []bigip.PolicyRuleCondition) []any {
+	att := make([]any, len(conditions))
 	for x, a := range conditions {
 		att[x] = DatainterfaceToResourceData(a)
 	}
 	return att
 }
 
-func DatainterfaceToResourceData(a interface{}) map[string]interface{} {
-	obj := make(map[string]interface{})
+func DatainterfaceToResourceData(a any) map[string]any {
+	obj := make(map[string]any)
 	v := reflect.ValueOf(a)
 	for i := 0; i < v.NumField(); i++ {
 		fn := toSnakeCase(v.Type().Field(i).Name)

--- a/bigip/datasource_bigip_ltm_policy_test.go
+++ b/bigip/datasource_bigip_ltm_policy_test.go
@@ -34,3 +34,53 @@ data "bigip_ltm_policy" "test" {
 }
 `, name)
 }
+
+// add test for subfolder policy
+func TestAccDataSourceBigipLtmPolicy_subfolder(t *testing.T) {
+	policyName := "/Common/folder1/ecopolicy"
+	resource.Test(t, resource.TestCase{
+		PreCheck:  func() { testAcctPreCheck(t) },
+		Providers: testAccProviders,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccDataSourceBigipLtmPolicySubfolderConfig(policyName),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("data.bigip_ltm_policy.test2", "name", policyName),
+				),
+			},
+		},
+	})
+}
+
+func testAccDataSourceBigipLtmPolicySubfolderConfig(name string) string {
+	return fmt.Sprintf(`
+data "bigip_ltm_policy" "test2" {
+  name = "%s"
+}
+`, name)
+}
+
+// add test for subfolder policy with different name
+func TestAccDataSourceBigipLtmPolicy_subfolder2(t *testing.T) {
+	policyName := "/Common/testpolicy"
+	resource.Test(t, resource.TestCase{
+		PreCheck:  func() { testAcctPreCheck(t) },
+		Providers: testAccProviders,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccDataSourceBigipLtmPolicySubfolder2Config(policyName),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("data.bigip_ltm_policy.test2", "name", policyName),
+				),
+			},
+		},
+	})
+}
+
+func testAccDataSourceBigipLtmPolicySubfolder2Config(name string) string {
+	return fmt.Sprintf(`
+data "bigip_ltm_policy" "test2" {
+  name = "%s"
+}
+`, name)
+}

--- a/bigip/version.go
+++ b/bigip/version.go
@@ -1,4 +1,4 @@
 package bigip
 
 // ProviderVersion is set at build-time in the release process
-var ProviderVersion = "1.23.0"
+var ProviderVersion = "1.24.0"

--- a/go.mod
+++ b/go.mod
@@ -10,8 +10,8 @@ require (
 	github.com/Azure/azure-storage-blob-go v0.13.0
 	github.com/Azure/go-autorest/autorest v0.11.18
 	github.com/Azure/go-autorest/autorest/adal v0.9.13
-	github.com/f5devcentral/go-bigip v0.0.0-20250808131030-ef245d09723a
-	github.com/f5devcentral/go-bigip/f5teem v0.0.0-20250808131030-ef245d09723a
+	github.com/f5devcentral/go-bigip v0.0.0-20250812165400-06e558f89430
+	github.com/f5devcentral/go-bigip/f5teem v0.0.0-20250812165400-06e558f89430
 	github.com/google/uuid v1.3.0
 	github.com/hashicorp/go-cty v1.4.1-0.20200414143053-d3edf31b6320
 	github.com/hashicorp/terraform-plugin-sdk/v2 v2.25.0

